### ISSUE TITLE
make pytorch internals const correct

### ISSUE THREAD

**Issue by dagitses** [pytorch/pytorch#97856]
### 🚀 The feature, motivation and pitch

The layers of tensor data throughout our stack have historically been lax about const correctness. We don't, for example, consistently distinguish between a data accessor granting mutable vs const access. This makes it challenging to add instrumentation that can distinguish data reads from data writes.

Furthermore, we will mark methods const that give out mutable handles to class internals.

We should be strict on our tensor class and its internals about these accesses.

There is a straightforward migration path that forces us to make a decision at each callsite: change the existing method to return a const pointer and introduce an equivalent variant named mutable_<original_method_name> that gives a non-const pointer. Then we just fix what breaks.

While doing this, we need to be careful about not just blindly changing to the mutable variant: we should only do so when we must.

Challenges:
Some of our libraries were designed without considering this and will require some redesign to accommodate this distinction. For example, TensorIterator levels all of its arguments to non-const. JIT function calls do the same.

### Additional context

We want this right now to instrument reads/writes with high fidelity to accurately simulate and also implement copy-on-write tensors.

**Comment by ezyang** [comment]
In this issue you seem to suggest preventing const references to Tensor from vending mutable data pointer. This is unworkable. A Tensor is a shared reference to a *mutable* tensor, const Tensor& says you have a const reference to a mutable tensor. We would have to introduce an entirely new type ConstTensor to do this. I recommend sticking to just making data produce const pointer and mutable_data non const

**Comment by ezyang** [comment]
IMO from_blob should just always materialize, *even* if we had ConstTensor

**Comment by kurtamohler** [comment]
I'd like to write down all the methods that I think need to be fixed for this issue.

`TensorBase::data_ptr()` was already marked for this issue in the codebase

https://github.com/pytorch/pytorch/blob/4916a7e94f44daafecd4ed7bec70a7b799a2a28b/aten/src/ATen/core/TensorBase.h#L582-L588

https://github.com/pytorch/pytorch/blob/4916a7e94f44daafecd4ed7bec70a7b799a2a28b/aten/src/ATen/core/TensorBase.h#L602-L606

Similarly, `DataPtr::get()` is a const-method, but it returns a non-const pointer. It should be changed to return a const pointer so that the caller cannot change the internal data. `DataPtr::mutable_get()` already exists, so that can be used when the mutable pointer is needed.

https://github.com/pytorch/pytorch/blob/e8e3afb784f28562aa9463da06c38b0fb574b01c/c10/core/Allocator.h#L38-L40

Similar story with `DataPtr::get_context()`, except that `DataPtr::mutable_get_context()` does not already exist and will need to be added.

https://github.com/pytorch/pytorch/blob/e8e3afb784f28562aa9463da06c38b0fb574b01c/c10/core/Allocator.h#L44-L46

Currently, it's possible to obtain a non-const pointer to a tensor's data using `tensor.storage().data_ptr().get()`, without causing a COW materialization to occur, when using the changes from PR #113397. `Storage::data_ptr()` correctly returns a `const DataPtr&` without materializing, but then `DataPtr::get()` gives a non-const pointer, as mentioned above.

After #113397 is merged, I will focus on fixing these. Is that alright with you @ezyang ?

**Comment by ezyang** [comment]
I have a bunch of in progress commits from @dagitses which may be helpful. They're currently only available internally but I can spend effort exporting them if you think it would help. Here's the list:

```
  o  bec99de1c  Tuesday at 15:02  mikeyd
  │  make magmaLdlHermitian data_ptr-correct
  │
  o  6c2cc7bfd  Tuesday at 15:02  mikeyd
  │  use mutable_data_ptr after calls to at::empty in BatchLinearAlgebra
  │
  o  cd8a6eb17  Tuesday at 15:02  mikeyd
  │  migrate cuda files to const_data_ptr
  │
  o  cf2e543c1  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/TensorTransformations.cu data_ptr-correct
  │
  o  0d07a4114  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/UpSampleTrilinear3d.cu data_ptr-correct
  │
  o  d7c9b9fb0  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/Col2Im.cu data_ptr-correct
  │
  o  a209df082  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/NaiveConvolutionTranspose3d.cu data_ptr-correct
  │
  o  09886c863  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/NaiveDilatedConvolution.cu data_ptr-correct
  │
  o  c11b8bdca  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/RangeFactories.cu data_ptr-correct
  │
  o  ce709704f  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/Shape.cu data_ptr-correct
  │
  o  63c4321b2  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/TensorFactories.cu data_ptr-correct
  │
  o  939f4490c  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/RreluWithNoise.cu
  │
  o  5f740a510  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/SortStable.cu
  │
  o  f4e8f9c74  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/AveragePool2d.cu data_ptr-correct
  │
  o  48479ba75  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/Bucketization.cu data_ptr-correct
  │
  o  7a3e8053f  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/Nonzero.cu data_ptr-correct
  │
  o  ecd9972bb  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/TensorModeKernel.cu data_ptr-correct
  │
  o  91aa86643  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/UpSampleNearest1d.cu data_ptr-correct
  │
  o  ea92ee7f8  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/UpSampleNearest3d.cu data_ptr-correct
  │
  o  523a08f08  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/DistanceKernel.cu data_ptr-correct
  │
  o  4258ca786  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/ReflectionPad.cu data_ptr-correct
  │
  o  792331e5f  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/AdaptiveAveragePooling3d.cu data_ptr-correct
  │
  o  1aaf80c9a  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/ConvolutionMM2d.cu data_ptr-correct
  │
  o  a75c227ce  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/DilatedMaxPool2d.cu data_ptr-correct
  │
  o  13051a563  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/DilatedMaxPool3d.cu data_ptr-correct
  │
  o  2528972f8  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/ForeachReduceOp.cu data_ptr-correct
  │
  o  b3cce94f8  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/UpSampleBilinear2d.cu data_ptr-correct
  │
  o  574f4391a  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/AmpKernels.cu data_ptr-correct
  │
  o  649cd371c  Tuesday at 15:02  mikeyd
  │  add comment indicating that `maskPrefixSum` is mutated
  │
  o  32e22d01f  Tuesday at 15:02  mikeyd
  │  make ATen/native/cuda/AdaptiveAveragePooling.cu data_ptr-correct
  │
  o  475268884  Tuesday at 15:02  mikeyd
  │  col2im
  │
  o  53b8f219f  Tuesday at 15:02  mikeyd
  │  im2col
  │
  o  28cc69ade  Tuesday at 15:02  mikeyd
  │  ATen/native/cuda/EmbeddingBag.cu
  │
  o  5385b571a  Tuesday at 15:02  mikeyd
╭─╯  ATen/native/cuda/ScanUtils.cuh
│
o  036833b7b  Tuesday at 15:01  remote/master
╷
o  a38c73c7e  Tuesday at 04:27  remote/fbsource/stable
╷
o  3803a2dba  Apr 14 at 11:01  remote/fbcode/stable
╷
o  a01f48e06  Apr 14 at 09:25  remote/fbandroid/stable remote/fbobjc/stable
╷
o  0ab0d1c49  Apr 14 at 08:00  remote/fbcode/warm
╷
╷ o  7b8a2fc6a  Apr 12 at 17:43  mikeyd
╷ │  lapackSvd
╷ │
╷ o  34eba4a71  Apr 12 at 17:43  mikeyd
╷ │  dot_impl
╷ │
╷ o  c2de6bb25  Apr 12 at 17:43  mikeyd
╷ │  lapackLdl
╷ │
╷ o  fa8877a2d  Apr 12 at 17:43  mikeyd
╷ │  lapackLdlSolve
╷ │
╷ o  b6d00fc34  Apr 12 at 17:43  mikeyd
╷ │  ormqr
╷ │
╷ o  570cff7fe  Apr 12 at 17:43  mikeyd
╷ │  blasTriangularSolve
╷ │
╷ o  a4961dfd9  Apr 12 at 17:43  mikeyd
╷ │  lapackOrmqr
╷ │
╷ o  1452c1b1b  Apr 12 at 17:43  mikeyd
╷ │  batch_iterator_with_broadcasting
╷ │
╷ o  339229e4c  Apr 12 at 17:43  mikeyd
╷ │  lapackGeqrf
╷ │
╷ o  ea56d8791  Apr 12 at 17:43  mikeyd
╷ │  axpy
╷ │
╷ o  a70988356  Apr 12 at 17:43  mikeyd
╷ │  lapackOrggr
╷ │
╷ o  6de681820  Apr 12 at 17:43  mikeyd
╷ │  lapackOrgqr
╷ │
╷ o  a7edfb090  Apr 12 at 17:43  mikeyd
╷ │  lapackSyevd
╷ │
╷ o  3ed381a7a  Apr 12 at 17:43  mikeyd
╷ │  apply_reflect_conj_tri_single
╷ │
╷ o  7a1b88551  Apr 12 at 17:43  mikeyd
╷ │  sparse_mm
╷ │
╷ o  c4ebb8eee  Apr 12 at 17:43  mikeyd
╷ │  gelsd
╷ │
╷ o  4f0c2f492  Apr 12 at 17:43  mikeyd
╷ │  lapackeig
╷ │
╷ o  0eeef5f80  Apr 12 at 17:43  mikeyd
╷ │  cholesky solve
╷ │
╷ o  c450cc463  Apr 12 at 17:43  mikeyd
╷ │  lapackEig
╷ │
╷ o  cdae2f651  Apr 12 at 17:43  mikeyd
╷ │  gemm
╷ │
╷ o  b2fbfc8aa  Apr 12 at 17:43  mikeyd
╷ │  lapackLuSolve
╷ │
╷ o  afd03fef7  Apr 12 at 17:43  mikeyd
╷ │  blak
╷ │
╷ o  9f747e120  Apr 12 at 17:43  mikeyd
╷ │  blak
╷ │
╷ o  e8481f7de  Apr 12 at 17:43  mikeyd
╷ │  owned or borrowed vector
╷ │
╷ o  6468fcf4a  Apr 12 at 17:43  mikeyd
╷ │  non cont
╷ │
╷ o  ae03afb76  Apr 12 at 17:43  mikeyd
╷ │  conditional_data_ptr
╷ │
╷ o  cc8634bb3  Apr 12 at 17:43  mikeyd
╷ │  qlinear deserialize
╷ │
╷ o  06e4f3b78  Apr 12 at 17:43  mikeyd
╷ │  cholesky inverse
╷ │
╷ o  8b018fee3  Apr 12 at 17:43  mikeyd
╷ │  cholesky
╷ │
╷ o  4076160c3  Apr 12 at 17:43  mikeyd
╷ │  repeat
╷ │
╷ o  3cbff7163  Apr 12 at 17:43  mikeyd
╷ │  blas vdot
╷ │
╷ o  d42a354c0  Apr 12 at 17:43  mikeyd
╷ │  blas dot
╷ │
╷ o  c32f90054  Apr 12 at 17:43  mikeyd
╷ │  gemv?
╷ │
╷ o  24262a091  Apr 12 at 17:43  mikeyd
╷ │  non cont
╷ │
╷ o  4d5f9561d  Apr 12 at 17:43  mikeyd
╷ │  use as_std_complex in BatchLinearAlgebra
╷ │
╷ o  f2f5d30fb  Apr 12 at 17:43  mikeyd
╷ │  use as_std_complex in blaskernel
╷ │
╷ o  1065a7164  Apr 12 at 17:43  mikeyd
╷ │  as_mutable
╷ │
╷ o  53996de08  Apr 12 at 17:43  mikeyd
╷ │  as_std_complex
╷ │
╷ o  7f6efb9cc  Apr 12 at 17:43  mikeyd  D44537619
╷ │  make TensorBase::data<T>() return a const pointer
╷ │
╷ o  d3d1954f8  Apr 12 at 17:43  mikeyd  D44537620
╷ │  dubious TensorBase::mutable_data<> changes
╷ │
╷ o  e53ff5a18  Apr 12 at 17:43  mikeyd  D44569821
╷ │  Copykernel
╷ │
╷ o  d646fddd2  Apr 12 at 17:43  mikeyd  D44492540
╷ │  make TensorBase::data_ptr() return a const pointer
╷ │
╷ o  50f7ade9b  Apr 12 at 17:43  mikeyd  D44492539
╷ │  TODO audit TensorBase::mutable_data_ptr()
╷ │
╷ o  0236350f9  Apr 12 at 17:43  mikeyd
╭─╯  TODO audit TensorBase::mutable_data_ptr<T>
│
│ o  720b04e3f  Apr 12 at 17:39  mikeyd  D44611380
├─╯  introduce TensorIterator::unsafe_replace_input()
│
o  1a1816363  Apr 12 at 17:38
╷
╷ o  2337f1410  Apr 12 at 12:43  mikeyd  D44836573
╭─╯  trivially migrate std::copy destinations to mutable_data_ptr
│
│ o  0a6c68b27  Apr 12 at 12:41  mikeyd  D44836571
├─╯  migrate memcpy src to const_data_ptr
│
o  6c37733d2  Apr 12 at 12:04
╷
╷ o  5f2299b7a  Apr 12 at 15:40  mikeyd  workspace
╭─╯  do not push WORKSPACE
│
o  e12b13fe3  Apr 12 at 08:07
╷
╷ o  793b8cddb  Apr 12 at 15:44  mikeyd  D44809218
╷ │  extract out c10/macros library in xplat and arvr
╷ │
╷ o  4cca8c2b8  Apr 12 at 15:44  mikeyd  D44809223
╷ │  add test to verify the correct preprocessor definitions in arvr
╷ │
╷ o  0b5ed1e86  Apr 12 at 15:44  mikeyd  D44775027
╷ │  add test to verify the correct preprocessor definitions in xplat
╷ │
╷ o  ceaf88dd9  Apr 12 at 15:44  mikeyd  D44656590
╷ │  convert cmake_configure_file into a rule
╷ │
╷ o  997d8b4b8  Apr 12 at 15:44  mikeyd  D44823439
╷ │  use exec for sed command in cmake_configure_file
╷ │
╷ o  66eb4ece4  Apr 12 at 15:44  mikeyd  D44809219
╷ │  move cmake configure_file emulation into a python file
╷ │
╷ o  4f077851c  Apr 12 at 15:44  mikeyd  D44910082
╷ │  use a hermetic Python toolchain in Bazel
╷ │
╷ o  03b02a67d  Apr 12 at 16:23  mikeyd  D44938260
╷ │  print stdout and stderr if bazel shasum linter fails
╷ │
╷ o  73955c9cd  Apr 12 at 10:56  mikeyd  D44919803
╷ │  use Bazelisk to fetch Bazel in CI
╷ │
╷ o  1a7d9080e  Apr 12 at 10:56  mikeyd  D44910081
╷ │  upgrade and use rules_python rather than built-in Python rules
╷ │
╷ o  868506857  Apr 12 at 10:56  mikeyd  D44922363
╷ │  run buildifier on the WORKSPACE
╷ │
╷ o  05ca01f36  Apr 12 at 03:08  mikeyd  D44788787
╷ │  stop implementing cmake configure_file emulation with a pipeline
╷ │
╷ o  15a492a29  Apr 12 at 03:08  mikeyd  D44788786
╷ │  extract the cmake_configure_file command generation into its own library
╷ │
╷ o  7cad8eaba  Apr 12 at 03:08  mikeyd  D44809224
╷ │  create a subdirectory for the cmake_configure_file rule implementation
╷ │
╷ o  5d53ce3f1  Apr 12 at 03:08  mikeyd  D44809222
╷ │  add test that bazel generates cmake macros as expected
╷ │
╷ o  32053f51e  Apr 12 at 03:08  mikeyd  D44809225
╷ │  generalize fbcode cmake macros test to work for any build system
╷ │
╷ o  d75f35dc1  Apr 12 at 03:08  mikeyd  D44823435
╷ │  extract out an _update_common_args and use it in the remaining rules
╷ │
╷ o  b08ac83ab  Apr 12 at 03:08  mikeyd  D44823438
╷ │  separate external deps processing from regular deps
╷ │
╷ o  3019d0b6d  Apr 12 at 03:08  mikeyd  D44823434
╷ │  extract common c++ translation to a shared function
╷ │
╷ o  f512e8271  Apr 12 at 03:08  mikeyd  D44823441
╷ │  extract out a function to rewrite visibility in-place
```

**Comment by ezyang** [comment]
Re the methods in question, I think the DataPtr method (and esp `get_context`) are lower priority, because they just don't have that much direct access.

**Comment by kurtamohler** [comment]
I figure that making TensorIterator const correct would have a relatively high impact since so many operators use it. So I'm going to focus on doing that first. It will require adding a new public API `TensorIteratorConfig::add_const_input`

**Comment by dagitses** [comment]
Take a look at https://github.com/pytorch/pytorch/pull/98162 which was intended to change the internals to support a distinction between mutable and immutable tensor arguments. This could be a useful precursor to what you're looking to add.

Are inputs always logically const in TensorIterator?

**Comment by kurtamohler** [comment]
Thanks for pointing me to that @dagitses!

> Are inputs always logically const in TensorIterator?

Currently no, since a TensorIterator loop function is given non-const pointers to both the outputs and inputs:

https://github.com/pytorch/pytorch/blob/8ddc549c0fba855cf734df9fe9e5fbd4b41e069b/aten/src/ATen/TensorIterator.h#L253-L254

Just to give an example, a user of the TensorIterator interface could do this:

```cpp
  at::Tensor out = at::empty({10});
  at::Tensor a = at::randn({10});
  at::TensorIteratorConfig iter_config;
  iter_config
    .add_output(out)
    .add_input(a);
  auto iter = iter_config.build();

  auto my_loop = [](char** data, const int64_t* strides, int64_t n) {
    float* output_data = reinterpret_cast<float*>(data[0]);
    float* input_data = reinterpret_cast<float*>(data[1]);

    // Of course we can write to the output, as we should be able to do
    *output_data = 0.123f;

    // But we can also write to the input
    *input_data = 0.456f;
  };

  iter.for_each(my_loop);
```

Note that even if we had something like `TensorIteratorConfig::add_const_input` or `TensorIterator::unsafe_replace_input`, it would avoid the COW materialization if we use `Tensor::const_data_ptr` but then constness gets casted away when it's given to the loop function.

So to really make TensorIterator const-correct, the loop function signature will need to be changed to separate const and non-const data pointers, something like this:

```cpp
  using loop2d_t = c10::function_ref<
      void(char** mutable_data, const int64_t* mutable_strides,
           const char** const_data, const int64_t* const_strides,
           int64_t size0, int64_t size1)>;
```

**Comment by kurtamohler** [comment]
@peterbell10 and @malfet, I see both of you have worked on TensorIterator, so I wanted to loop you into this conversation. I'm curious to know what you think about my previous comment.

To summarize: I want to make TensorIterator avoid causing a [COW tensor](https://github.com/pytorch/pytorch/issues/109833) to be materialized if the tensor is only going to be read from, not written to. I think that one of things I'll need to do is split the readonly data out into its own `const char**` argument in the TensorIterator loop function. I'd like to make sure this is the right idea and that it's alright to change the TensorIterator API. Or if there's a better option, let me know

**Comment by peterbell10** [comment]
I would be wary about bc-breaking changes to the `TensorIterator ` API as it would break extensions, which we have no control over. I think it should be possible to add a new overload and migrate our code to use it though.

**Comment by ezyang** [comment]
I think BC breaking to TI is less of a problem in the world of PT2.0. Most people who want speedy pointwise/reductions are going to use inductor instead. That being said, there are a lot of call sites in PT proper you'll have to fix when you BC break. But I think it can be done.

**Comment by kurtamohler** [comment]
Despite the fact that we've removed all the unnecessary materializations in the forward and backward calls for all the operators in the `op_db` list, there are still some unnecessary materializations elsewhere, which I've found just by grepping for calls to `data_ptr`. I will submit a series of PRs to update them to use `const_data_ptr`

**Comment by kurtamohler** [comment]
Oh actually, I found about 200 cases with `git grep '^\s*\<const\>.*\<data_ptr\>'`, which can be automatically replaced
